chore: open firewall ports for SSH server  proxy

kian99 opened the pull request:
This PR is a draft while I gather confirmation on the correct way to do this.

This PR is part of the SSH proxy work. Here, we are looking to open the `ssh-server-port` on controller machines, for all providers. Note that this port is distinct from port 22 used to expose the controller's sshd daemon, providing ssh access to the machine where the jujud agent runs. Intead, this port (default 17022) is the one that jujud binds to, providing an ssh server and acting as a proxy for SSH access to units.

In this PR I've updated the following providers:
- Azure
- Common
- GCE
- OCI
- Equinix (I believe this is deprecated but I've updated it anyway)
- Kubernetes

I've also updated the firewaller's `ModelFirewallRules` method to ensure the port remains open on changes. The ssh-server port is a read/write controller config value. So a follow-up PR will be needed to watch for changes and apply them.

**NB::** We should also consider not merging this change to avoid opening the ssh-server port until the remaining work is done.

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [x] Go unit tests, with comments saying what you're testing
- [ ] [Integration tests](https://github.com/juju/juju/tree/main/tests), with comments saying what you're testing
- [ ] [doc.go](https://discourse.charmhub.io/t/readme-in-packages/451) added or updated in changed packages

## QA steps

I've only tested the changes for Azure as I don't have credentials for the other clouds.

1. `make install`
2. `juju bootstrap azure ssh-server-test --build-agent`
3. `juju show-controller`, copy the public IP address.
4. `nc -vz <ip> 17022`
`Connection to <ip> 17022 port [tcp/*] succeeded!`

## Links

<!-- Link to all relevant specification, documentation, bug, issue or JIRA card. -->

**Launchpad bug:** https://bugs.launchpad.net/juju/+bug/

**Jira card:** [JUJU-7360](https://warthogs.atlassian.net/browse/JUJU-7360)

wallyworld commented:
We really want to avoid hard coding ports like this. We did it this way dating back to 2016 (or earlier) for the controller api port, and again in 2018 for the let's encrypt port. That's when we had no other alternative.

Now however, we have the controller charm which gives us a controller app. We can use standard Juju modelling for the ssh server port. Doing it manually, it's:

```
bootstrap a vanilla 3.6 controller, no code changes
juju exec -u controller/0 -- open-port 17022
juju expose controller
```

And then it even shows up correctly in status etc. Note it is marked as `exposed`.
```

App         Version  Status  Scale  Charm            Channel     Rev  Exposed  Message
controller           active      1  juju-controller  3.6/stable  116  yes      

Unit           Workload  Agent  Machine  Public address  Ports      Message
controller/0*  active    idle   0        128.203.173.11  17022/tcp  
```

So what we want to do is:
- when the controller app is created at bootstrap, set exposed = true
- whenever a new controller unit is added, open the ssh server port on that unit
- disallow the ssh server port from being closed (in facade and state apis)

The above are super simple.

Bonus points in a follow up PR - remove the existing hard coding for ports 17070 and 80 in the providers and do the above for those as well.

kian99 commented:
> * when the controller app is created at bootstrap, set exposed = true
> * whenever a new controller unit is added, open the ssh server port on that unit
> * disallow the ssh server port from being closed (in facade and state apis)

For the first two steps here, are you suggesting we do this from the Jujud code? And if so, how can/should we interact with the charm from the agent binary?

wallyworld commented:
> > * when the controller app is created at bootstrap, set exposed = true
> > * whenever a new controller unit is added, open the ssh server port on that unit
> > * disallow the ssh server port from being closed (in facade and state apis)
> 
> For the first two steps here, are you suggesting we do this from the Jujud code? And if so, how can/should we interact with the charm from the agent binary?

We don't interact with the charm - my manual steps were an example of how to show the mechanics of it.
The jujud code updates the juju model. It sets exposed=true on the app and updates the model to record the open port the same as if the charm had called open-port.

kian99 commented on state/enableha.go:
```diff
@@ -153,9 +154,45 @@ func (st *State) EnableHA(
 	}
+	// Failing to open the SSH proxy port shouldn't block enabling HA.
+	// The user can always open the port manually.
+	if err := st.openSSHProxyPort(); err != nil {
```
Note, this function is not inside the above transaction because the primitives for opening SSH ports assume we have a `unit` object, which isn't available until after the transaction executes. I think it's probably more work than its worth to write the raw documents for opening the SSH ports in the transaction above.

wallyworld replied:
We want to add the ops to open the port in the `enableHAIntentionOps` method. See the code for adding the unit below - we want to add to that. We already have the means to get the ops for opening a port.
```
		if controllerApp != nil {
			_, unitOps, err := controllerApp.addUnitOps("", AddUnitParams{
				UnitName:  &controllerUnitName,
				machineID: mdoc.Id,
			}, nil)
			if err != nil {
				return nil, ControllersChanges{}, errors.Trace(err)
			}
			ops = append(ops, unitOps...)
		}
```

wallyworld replied:
I push a change to make this work https://github.com/juju/juju/pull/19180/commits/4ae859af6393b5f9b176669f51522c22e82e9c9c

I tested with enable-ha to new machines - we also need to verify that it works with ha placement where the machine exists and we use --to N to convert that machine to a controller. I suspect the added logic will need to be tweaked slightly since the machine exists but not the unit. So we'll need another pre-check. But you get the idea.

kian99 commented on cmd/jujud/agent/controllercharm.go:
```diff
@@ -342,5 +408,16 @@ func addControllerApplication(
 		return nil, errors.Trace(err)
 	}
+
+	// Expose all endpoints of the controller app
```
Should we be exposing all endpoints of the controller app?

wallyworld replied:
We don't have an ssh endpoint so we have no control via spaces over what CIDRs to use so for now we can start with `0.0.0.0` but should consider tightening this up as part of the work.

wallyworld commented on caas/kubernetes/provider/bootstrap.go:
```diff
@@ -644,6 +646,11 @@ func (c *controllerStack) createControllerService(ctx context.Context) error {
 					TargetPort: intstr.FromInt(c.portAPIServer),
 					Port:       int32(c.portAPIServer),
 				},
+				{
+					Name:       "ssh-server",
+					TargetPort: intstr.FromInt(c.portSSHServer),
+					Port:       int32(c.portSSHServer),
+				},
```
I'd like to understand why we need this. The k8s firewaller should do the right thing. Yet clearly it doesn't cause I can manually open another port eg17666 and the db is correct
```
  unit-port-ranges:
    controller/0:
      "":
      - fromport: 17022
        protocol: tcp
        toport: 17022
      - fromport: 17666
        protocol: tcp
        toport: 17666

```
yet the ports aren't opened. We either have a bug or there's something we didn't finish. There was a task a while back to do this so we'll need to see what happened.

kian99 replied:
I took a look at how k8s charms open ports, and saw that charms are making use of the `lightkube.models.core_v1` library to mutate k8s objects directly, one example [here](https://github.com/canonical/openfga-operator/blob/main/src/charm.py#L39C6-L39C30). I recall discussion that Juju and k8s networking was not something that was fleshed out, so I assume there is no mechanism to open a k8s unit port entirely from Juju.

With that in mind, should I leave the changes I've made to the k8s provider? I think we'll need to discuss the k8s side of this all separately. I know currently the Juju CLI will port forward the Juju controller so that it's available outside the cluster, for SSH we could do something similar but then you wouldn't be able to use a plain openSSH client unless you setup an ingress/load-balancer.

wallyworld commented on cmd/jujud/agent/controllercharm.go:
```diff
@@ -34,95 +35,160 @@ import (
+		dataDir:                   c.DataDir(),
 	}
 
+	if err := ccd.getControllerDetails(st); err != nil {
+		return err
+	}
+
+	if err := ccd.deployCharm(st); err != nil {
+		return err
+	}
+
+	if err := ccd.finishUnitSetup(st); err != nil {
+		return err
+	}
+
+	if err := ccd.openProxySSHPort(st); err != nil {
+		return err
+	}
```
Just have a single `deploy()` method which does this in its implementation

wallyworld reviewed (commented):
Looking much better with the model driven approach.
The main thing to fix is the enable ha code to use txn ops.
QA passes after bootstrap but enable-ha resulted in the machines not starting and the firewaller bouncing.

Also, seems like an existing issue, but I'd like to understand the k8s behaviour. Will look into it.

kian99 commented:
> QA passes after bootstrap but enable-ha resulted in the machines not starting and the firewaller bouncing.

That's strange, I tested it on LXD and it worked. I didn't even add any code that should cause enable-ha to fail, if opening the ssh port fails, it just logs it and continues without error.

What provider did you test on?

wallyworld commented:
> > QA passes after bootstrap but enable-ha resulted in the machines not starting and the firewaller bouncing.
> 
> That's strange, I tested it on LXD and it worked. I didn't even add any code that should cause enable-ha to fail, if opening the ssh port fails, it just logs it and continues without error.
> 
> @wallyworld What provider did you test on?

Tested on Azure. LXD doesn't support OpenPorts et al so you need to test on a provider like AWS or Openstack or Azure etc that support it.

kian99 commented:
@wallyworld 
There are roughly 73 tests failing with the same error,
>value *params.Error = &params.Error{Message: "getting controller application: application \"controller\" not found", Code: "not found"} ("getting controller application: application \"controller\" not found")

This is due to the change where you've removed the `if controllerApp != nil {` checks and I presume these tests are not running the full bootstrap logic which deploys the controller charm. So I'll put the `if controllerApp != nil {` checks back in.

kian99 commented:
Last failing test is `TestEnableHAAddsControllerCharmToPromoted`, which is resulting in `transaction aborted`. @wallyworld any idea how I can dig into what is causing the transaction to be aborted?

SimoneDutto commented on state/enableha_test.go:
```diff
@@ -672,3 +673,38 @@ func (s *EnableHASuite) TestRemoveControllerMachineRace(c *gc.C) {
+
+		foundSSHProxyPort := false
+		for _, portRange := range openPorts {
+			if portRange.FromPort <= expectedPort && portRange.ToPort >= expectedPort {
```
why is it `>=` and `<=`?

wallyworld commented:
> @wallyworld There are roughly 73 tests failing with the same error,
> 
> > value *params.Error = &params.Error{Message: "getting controller application: application "controller" not found", Code: "not found"} ("getting controller application: application "controller" not found")
> 
> This is due to the change where you've removed the `if controllerApp != nil {` checks and I presume these tests are not running the full bootstrap logic which deploys the controller charm. So I'll put the `if controllerApp != nil {` checks back in.

The tests are faulty in that case. We should always expect the controller app to be there as it's created at bootstrap and should fail the workflow if there's an issue. If it's too hard to fix the tests, we can leave things as there were.

wallyworld commented:
> > QA passes after bootstrap but enable-ha resulted in the machines not starting and the firewaller bouncing.
> 
> That's strange, I tested it on LXD and it worked. I didn't even add any code that should cause enable-ha to fail, if opening the ssh port fails, it just logs it and continues without error.
> 
> @wallyworld What provider did you test on?

lxd doesn't really use the firewaller. 
I tested on Azure, but openstack or aws will do also.

SimoneDutto reviewed (approved):
lgtm, but since we are touching the bootstrap code i would be inclined to have 2 approvals from juju members

manadart commented on cmd/jujud/agent/controllercharm.go:
```diff
@@ -342,5 +418,16 @@ func addControllerApplication(
 	}
+
+	// Expose all endpoints of the controller app
+	mappedExposeParams := map[string]state.ExposedEndpoint{
```
Controllers on public have a public address. Doesn't this allow anyone to get to any of these endpoints?
```
provides:
  dashboard:
    interface: juju-dashboard
  website:
    interface: http
  metrics-endpoint:
    interface: prometheus_scrape
```

kian99 replied:
I've run the following on Azure with Juju 3.6.4,
```
$ juju bootstrap azure test-ssh
$ curl https://20.164.128.166:17070/introspection/metrics -k
authentication failed: no credentials provided
```
So the metrics endpoint is already exposed, but guarded by basic auth.
Looking at the [controller charm](https://github.com/juju/juju-controller/blob/3.6/src/charm.py), the other 2 endpoints don't seem to go to a specific path beyond the root and don't use any port beyond the API server's port.

manadart commented on cmd/jujud/agent/controllercharm.go:
```diff
@@ -34,95 +35,170 @@ import (
+	dataDir           string
+}
+
+func (c *BootstrapCommand) deployControllerCharm(st *state.State, args controllerCharmArgs) error {
```
Move this out from between your `controllerCharmDeployer` declarations. Makes for junky reading.

manadart reviewed (commented):
Not sure about the blanket exposure.

We should add bootstrap tests for the changed behaviour in cmd/juju/agent. Surely we can verify some of the new side-effects we've added here.

kian99 commented:
> We should add bootstrap tests for the changed behaviour in cmd/juju/agent. Surely we can verify some of the new side-effects we've added here.

Done, I was able to modify the `assertControllerApplication` used in the tests to assert that the controller app is exposed and that the correct port is exposed.

kian99 commented:
@wallyworld @manadart 
I have successfully QA'd this change on Azure. But when I run `juju enable-ha` the 2 additional machines are never provisioned. I have waited several hours and their status never changes from `pending` when I run `juju status`.
I tried this with Juju 3.6.4 and the issue is the same so I don't think this is due to anything in this PR. Am I missing something on how to enable HA before I open a ticket? I will test AWS too.

kian99 commented:
There is a subtle issue when enabling HA. It seems the firewaller [loop](https://github.com/juju/juju/blob/3.6/internal/worker/firewaller/firewaller.go#L327) is not being triggered after the enable HA operation, so although the new port intent is correct in the DB, nothing is updating the provider.
On AWS I verified this by
1. bootstrap
2. `juju enable-ha`
3. wait for new units to be ready
4. Check AWS dashboard and observe that units 1 and 2 do not have port 17022 open. Also verify with `nc -vz <ip> 17022`, only machine 0 has it opened.
5. Trying to restart jujud on machines on non-leader units does nothing.
6. Restart jujud on the leader unit, immediately the logs of the leader machine show that the ports have been opened.

It's surprising to me that [this machine watcher](https://github.com/juju/juju/blob/3.6/internal/worker/firewaller/firewaller.go#L347) is not getting triggered. My best guess is possibly the new controller machines are not provisioned yet when this fires and the ports can't be opened. Fixing this could be better left to a separate PR, open to ideas here.

manadart reviewed (commented):
I observed the same behaviour on AWS. 

The firewaller logic you indicated is not really the path that should be invoked here. When we add a machine, we will watch and add units. If those units change opened port ranges, the firewall acts.

The issue here is that the the firewaller assumes (correctly to-date) that because a unit is responsible for opening ports, then the machine will always be provisioned when it acts, because it couldn't have been deployed otherwise.

Now that we see it in theatre, I don't think special casing is the wisest course. I think we should have added this endpoint to the controller charm and made it optional for the user to expose the controller endpoint.

manadart reviewed (approved):
Right.

First of all, this isn't a `chore`. It adds functionality, so it's a `feat`.

Based on examination, the firewaller is incorrect, regardless of operating correctly to-date. This is based on three simple characteristics.
1. We watch machines for added/removed/life-changed.
2. We skip machine processing if the machine is not provisioned.
3. *But* we do not watch for the provisioned status to change (via instance data).

The follow-up will be to add `WatchInstanceData` to `startMachine`. This will necessitate a new method for the call on the firewaller facade.

kian99 commented:
@manadart I've updated one of the commits to `feat` over `chore`, updated the title of the PR and created [a card](https://warthogs.atlassian.net/browse/JUJU-7827) for the follow-up.

kian99 commented:
/merge
